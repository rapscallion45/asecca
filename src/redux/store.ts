import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
