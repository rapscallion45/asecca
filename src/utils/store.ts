/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
