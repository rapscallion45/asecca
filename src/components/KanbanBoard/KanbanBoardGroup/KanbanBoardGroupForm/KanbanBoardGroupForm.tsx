import { FC } from 'react';
import {
  Button,
  InputAdornment,
  CircularProgress,
  TextField,
} from '@mui/material';
import LabelImportantIcon from '@mui/icons-material/LabelImportant';
import { IEditKanbanBoardGroupPayload } from '@/redux/types';
import { IKanbanBoardColumn } from '@/lib/api/api-types';
import useKanbanBoardTaskFormController from './KanbanBoardGroupFormController';

/**
 * Kanban Board Group Form Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.9
 *
 * @typedef IKanbanBoardGroupFormProps
 * @prop {boolean} isEditMode - determines whether this is a new group or editing
 * @prop {boolean} canEdit - group can be edited
 * @prop {Array<IKanbanBoardColumn>} columns - columns for group's board
 * @prop {IEditKanbanBoardGroupPayload} editData - group data to be edited
 * @prop {any} closeModal - on close modal callback handler
 */
interface IKanbanBoardGroupFormProps {
  isEditMode: boolean;
  canEdit?: boolean;
  columns: Array<IKanbanBoardColumn>;
  editData?: IEditKanbanBoardGroupPayload;
  closeModal?: () => void;
}

/**
 * Kanban Board Group Form
 *
 * Kanban board group form interface for collecting user input when creating
 * or editing a group's details
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.9
 *
 * @component
 * @param {IKanbanBoardGroupFormProps} props - component props
 * @returns {FC} - kanaban board group form functional component
 */
const KanbanBoardTaskForm: FC<IKanbanBoardGroupFormProps> = (props) => {
  const { isEditMode, canEdit = false, columns, editData, closeModal } = props;
  const { saving, formik } = useKanbanBoardTaskFormController(
    isEditMode,
    columns,
    editData,
    closeModal
  );

  return (
    <form onSubmit={formik.handleSubmit}>
      <TextField
        fullWidth
        variant="outlined"
        margin="normal"
        id="name"
        name="name"
        label="Group Name"
        type="text"
        value={formik.values.name}
        onChange={formik.handleChange}
        error={formik.touched.name && Boolean(formik.errors.name)}
        helperText={formik.touched.name && formik.errors.name}
        autoComplete="on"
        disabled={!canEdit}
        InputProps={{
          startAdornment: (
            <InputAdornment position="start">
              <LabelImportantIcon color="secondary" />
            </InputAdornment>
          ),
        }}
      />
      {!isEditMode ? (
        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="secondary"
          disabled={saving || !canEdit}
          sx={{ padding: '10px 0', marginTop: '20px' }}
        >
          {!saving && 'Add Group'}
          {saving && <CircularProgress size={25} color="inherit" />}
        </Button>
      ) : (
        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="secondary"
          disabled={saving || !canEdit}
          sx={{ padding: '10px 0', marginTop: '20px' }}
        >
          {!saving && 'Update'}
          {saving && <CircularProgress size={25} color="inherit" />}
        </Button>
      )}
    </form>
  );
};

export default KanbanBoardTaskForm;
