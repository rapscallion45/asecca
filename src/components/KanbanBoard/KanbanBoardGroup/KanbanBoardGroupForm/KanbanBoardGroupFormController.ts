import { useDispatch } from 'react-redux';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import {
  IAddKanbanBoardGroupPayload,
  IEditKanbanBoardGroupPayload,
  IKanbanBoardState,
} from '@/redux/types';
import { AppDispatch } from '@/redux/store';
import {
  useSliceActions,
  useSliceSelector,
} from '@/components/SliceProvider/SliceProvider';
import { IKanbanBoardColumn } from '@/lib/api/api-types';

/**
 * Kanban board group form controller hook, used for group form logic,
 * input handling and form submission
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.9
 * @memberof KanbanBoardGroupForm
 *
 * @function
 * @param {boolean} isEditMode - is form creating a new group or editing existing
 * @param {Array<IKanbanBoardColumn>} columns - column list for groups's board
 * @param {IEditKanbanBoardGroupPayload} currentData - group data
 * @param {any} closeModal - callback for closing the group form modal
 */
const useKanbanBoardGroupFormController = (
  isEditMode: boolean,
  columns: Array<IKanbanBoardColumn>,
  currentData?: IEditKanbanBoardGroupPayload,
  closeModal?: () => void
) => {
  const dispatch = useDispatch<AppDispatch>();
  const { saving } = useSliceSelector() as IKanbanBoardState;
  const { addTask, editTask } = useSliceActions();

  /**
   * Yup input validation configuration for group form
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.9
   *
   * @constant
   */
  const validationSchema = Yup.object().shape({
    title: Yup.string()
      .required('Group name is required')
      .max(100, 'Group name must not exceed 100 characters'),
  });

  /**
   * Data submission handler for group form
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.9
   *
   * @method
   * @param {IEditKanbanBoardGroupPayload} payload - data to be submitted
   */
  const handleSubmit = (payload: IEditKanbanBoardGroupPayload) => {
    if (isEditMode) {
      // @ts-ignore
      dispatch(editTask(payload as IEditKanbanBoardGroupPayload));
      if (closeModal) closeModal();
    } else {
      // @ts-ignore
      dispatch(addTask(payload as IAddKanbanBoardGroupPayload));
      if (closeModal) closeModal();
    }
  };

  /**
   * Formik configuration for group form
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.9
   *
   * @constant
   */
  const formik = useFormik({
    initialValues: {
      name: currentData?.name || 'New Group',
      status: currentData?.status || '',
      groupIndex: currentData?.groupIndex || 0,
      newColIndex: columns.findIndex(
        (col: IKanbanBoardColumn) => col.name === currentData?.status
      ),
      prevColIndex: columns.findIndex(
        (col: IKanbanBoardColumn) => col.name === currentData?.status
      ),
    },
    validationSchema,
    onSubmit: (payload: IEditKanbanBoardGroupPayload) => {
      handleSubmit({
        ...payload,
        newColIndex: columns
          .map((col: IKanbanBoardColumn) => col.name)
          .indexOf(payload.status),
      });
    },
  });

  return { saving, formik };
};
export default useKanbanBoardGroupFormController;
