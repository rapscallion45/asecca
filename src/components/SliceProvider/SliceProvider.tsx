import { FC, ReactNode, createContext, useContext } from 'react';
import { useSelector } from 'react-redux';
import { Slice } from '@reduxjs/toolkit';
import { AppState } from '@/redux/store';

/**
 * Default value for the slice context
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof SliceProvider
 *
 * @constant
 * @type {Slice<Object>}
 */
const SliceContext = createContext<Slice>({} as Slice);

/**
 * Slice Provider Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef ISliceProviderProps
 * @prop {Slice} slice - context slice to provide down to child nodes
 * @prop {ReactNode} children - child nodes
 */
interface ISliceProviderProps {
  slice: Slice;
  children: ReactNode;
}

/**
 * Slice Provider
 *
 * Slice provider functional component wrapper for providing global state
 * managment functionality to child components
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @component
 * @param {ISliceProviderProps} props - component props
 * @returns {FC} - slice provider functional component
 */
const SliceProvider: FC<ISliceProviderProps> = (props) => {
  const { slice, children } = props;
  return (
    <SliceContext.Provider value={slice}>{children}</SliceContext.Provider>
  );
};

/**
 * Provider specific actions interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {CaseReducerActions<any, string>}
 */
export const useSliceActions = () => useContext<Slice>(SliceContext).actions;

/**
 * Provider specific selector interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {AppState}
 */
export const useSliceSelector = () => {
  const { name } = useContext<Slice>(SliceContext);
  return useSelector((state: AppState) => state[name as keyof AppState]);
};

export default SliceProvider;
