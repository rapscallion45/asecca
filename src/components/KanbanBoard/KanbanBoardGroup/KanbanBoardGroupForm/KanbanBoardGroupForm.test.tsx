import React from 'react';
import { Provider } from 'react-redux';
import {
  render,
  screen,
  waitFor,
  act,
  fireEvent,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '@/redux/store';
import SliceProvider from '@/components/SliceProvider/SliceProvider';
import KanbanBoardGroupForm from './KanbanBoardGroupForm';

/* default test query ID */
const query: string = '66135000001760012';

/* mock close modal callback */
const mockCloseModalCallback = jest.fn(() => {});

/**
 * Kanban Board Group Form Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board Group Form', () => {
  describe('Values', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render expected input fields and display values', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoardGroupForm
                columns={store.getState().collectionsKanban.data.columns}
                isEditMode
                canEdit
              />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      expect(screen.getByLabelText('Group Name')).toBeInTheDocument();
      expect(screen.getByText('Update')).toBeInTheDocument();
    });
  });

  describe('Editing and Saving', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render button loading spinner when saving data', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardGroupForm
              columns={store.getState().collectionsKanban.data.columns}
              isEditMode
              canEdit
              closeModal={mockCloseModalCallback}
            />
          </SliceProvider>
        </Provider>
      );

      /** Assert - wait for values to be loaded into form */
      expect(screen.getByLabelText('Group Name')).toBeInTheDocument();

      /** Act - update values */
      act(() => {
        fireEvent.change(screen.getByLabelText('Group Name'), {
          target: { value: 'Test Group' },
        });
      });

      /** Assert - wait for values to update */
      await waitFor(() => {
        expect(screen.getByDisplayValue('Test Group')).toBeInTheDocument();
      });

      /** Act - click Update button to trigger API save functionality */
      act(() => {
        fireEvent(
          screen.getByText('Update'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });

      /** Assert - Update button clicked and so no longer rendered */
      await waitFor(() => {
        expect(screen.queryByDisplayValue('Update')).toBeNull();
      });
    });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardGroupForm
              columns={store.getState().collectionsKanban.data.columns}
              isEditMode={false}
              canEdit
            />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
