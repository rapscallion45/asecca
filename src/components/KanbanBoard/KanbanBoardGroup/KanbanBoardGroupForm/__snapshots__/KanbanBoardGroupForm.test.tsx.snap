// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Group Form Renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="name"
      id="name-label"
    >
      Group Name
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-kw0y8f-MuiSvgIcon-root"
          data-testid="LabelImportantIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m3.5 18.99 11 .01c.67 0 1.27-.33 1.63-.84L20.5 12l-4.37-6.16c-.36-.51-.96-.84-1.63-.84l-11 .01L8.34 12 3.5 18.99z"
          />
        </svg>
      </div>
      <input
        aria-invalid={false}
        autoComplete="on"
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="name"
        name="name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="New Group"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Group Name
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1p5sqbq-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    Add Group
  </button>
</form>
`;
